src/internal/operations/backup.go

```diff
@@ -154,11 +163,147 @@ func (op *BackupOperation) Run(ctx context.Context) (err error) {
 	return err
 }
 
+// calls kopia to retrieve prior backup manifests, metadata collections to supply backup heuristics.
+func produceManifestsAndMetadata(
+	ctx context.Context,
+	kw *kopia.Wrapper,
+	sw *store.Wrapper,
+	sel selectors.Selector,
+	acct account.Account,
+) ([]*snapshot.Manifest, []data.Collection, error) {
+	complete, closer := observe.MessageWithCompletion("Fetching backup heuristics:")
+	defer func() {
+		complete <- struct{}{}
+		close(complete)
+		closer()
+	}()
+
+	m365, err := acct.M365Config()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		tid         = m365.AzureTenantID
+		oc          = selectorToOwnersCats(sel)
+		collections []data.Collection
+	)
+
+	ms, err := kw.FetchPrevSnapshotManifests(
+		ctx,
+		oc,
+		map[string]string{kopia.TagBackupCategory: ""})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, man := range ms {
+		if len(man.IncompleteReason) > 0 {
+			continue
+		}
+
+		bup := backup.Backup{}
+
+		if err := sw.Get(
+			ctx,
+			model.BackupSchema,
+			model.StableID(man.Tags[kopia.TagBackupID]),
+			&bup,
+		); err != nil {
+			return nil, nil, err
+		}
+
+		for _, fn := range graph.MetadataFileNames() {
```

Comment:
depending on how you want to handle errors (and logging) this whole loop block could be moved into `collectMetadata` if you wanted. `collectMetadata` could then generate all paths and then make a single call to `RestoreMultipleItems`. It would be a tad more efficient in the long run because it would only load the root directory of each snapshot once instead of once for every (resource owner, service) tuple. Not that there should be many tuples, but it plays a bit better with the other logic

Reply:
Good call, I've shifted that arround.